### PR TITLE
Feature save expanded story

### PR DESCRIPTION
**Pull request by kostadriano**
This PR is a feature to save a story in board V2 ref: [22729](https://central.cm42.io/projects/central-board-v2#story-22729)
- Refactor functions in reducer story to story model

### CODE REVIEW

File: app/assets/javascripts/actions/projectBoard.js [hunk 1]
```diff
@@ -24,7 +23,7 @@ const receiveProject = data => ({
 });
 
 export const fetchProjectBoard = projectId => {
-  return dispatch => {
+  return (dispatch , getState, { ProjectBoard }) => {
```
Comment (willamesoares) [inline app/assets/javascripts/actions/projectBoard.js]: remove space before first comma :nail_care:

File: app/assets/javascripts/models/beta/story.js [hunk 2]
```diff
@@ -60,3 +62,52 @@ export const isStoryNotEstimated = (storyType, estimate) => storyType === 'featu
...
+  };
+};
+
+export const editStory = (story, newAttributes) => {
+  const newStory = {
+    ...story._editing,
+    ...newAttributes
+  };
+  
+  newStory.estimate = !isFeature(newStory) ? null : newStory.estimate;
```
Comment (matheusazzi) [inline app/assets/javascripts/models/beta/story.js]: Maybe `newStory.estimate = isFeature(newStory) ? newStory.estimate : null;`

File: app/assets/javascripts/actions/projectBoard.js (hunk 1 above)
Comment (talyssonoc) [reply app/assets/javascripts/actions/projectBoard.js]: Hey @kostadriano, the space is still there 😛

File: app/assets/javascripts/actions/story.js [hunk 3]
```diff
@@ -10,6 +10,24 @@ export const toggleStory = (id) => ({
   id
 });
 
+export const updateStorySuccess = (story) => ({
+  type: actionTypes.UPDATE_STORY_SUCCESS,
+  story
+})
```
Comment (talyssonoc) [inline app/assets/javascripts/actions/story.js]: Missing a `;` here.

File: app/assets/javascripts/actions/story.js [hunk 4]
```diff
@@ -10,6 +10,24 @@ export const toggleStory = (id) => ({
...
+
+export const updateStory = (story, projectId) => {
+  return (dispatch , getState, { Story }) => {
```
Comment (talyssonoc) [inline app/assets/javascripts/actions/story.js]: There's an extra space here too.

File: app/assets/javascripts/actions/story.js [hunk 5]
```diff
@@ -10,6 +10,24 @@ export const toggleStory = (id) => ({
...
+    if (story._editing._isDirty) {
+      return Story.update(story._editing, projectId)
+        .then(( story ) => {
+          dispatch(updateStorySuccess(story));
+          dispatch(toggleStory(story.id));
```
Comment (talyssonoc) [inline app/assets/javascripts/actions/story.js]: _What if_ we add a `case` [in the reducer](https://github.com/Codeminer42/cm42-central/blob/master/app/assets/javascripts/reducers/stories.js#L10) to toggle the story when `UPDATE_STORY_SUCCESS` happens as well so we don't have to trigger two actions here? It'll make we save some processing time by not rendering the story twice.

File: app/assets/javascripts/reducers/stories.js [hunk 6]
```diff
@@ -8,12 +8,32 @@ const storiesReducer = (state = initialState, action) => {
     case actionTypes.RECEIVE_STORIES:
       return action.data;
     case actionTypes.TOGGLE_STORY:
-      return toggleStories(state, action.id);
+      return state.map(
+        updateIfSameId(action.id, (story) => {
```
Comment (talyssonoc) [inline app/assets/javascripts/reducers/stories.js]: You can make it shorter like this:

```js
updateIfSameId(action.id, toggleStory);
```

File: spec/javascripts/actions/story_spec.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  });
+
+  it('dispatch toggleStory and updateStorySuccess when _isDirty', (done) => {
+    const editedStory = {
+      ...story,
+      _editing: {
+        ...story,
+        _isDirty: true
+      }
+    }
```
Comment (talyssonoc) [inline spec/javascripts/actions/story_spec.js]: Missing a `;`

File: spec/javascripts/models/beta/story_spec.js [hunk 8]
```diff
@@ -210,4 +210,79 @@ describe('Story model', function() {
...
+      });
+    });
+
+    describe("When story is expanded", () => {
+      it("collapse story", () => {
+        const story = { collapsed: false }
+
+        const collapsedStory = Story.toggleStory(story);
+
+        expect(collapsedStory.collapsed).toEqual(true);
```
Comment (talyssonoc) [inline spec/javascripts/models/beta/story_spec.js]: Can we check the whole content of `collapsedStory` so we know `toggleStory` doesn't change more than the value of `collapsed`?

File: CHANGELOG.md [hunk 9]
```diff
@@ -4,10 +4,18 @@
 ### Added
 - Select to Story Type in board V2
 - Select to Story Estimate in board V2
+- Save option to Story in board V2
+- Package: es6-promise: 4.2.5
```
Comment (talyssonoc) [inline CHANGELOG.md]: I think we don't need to add that to the changelog.

File: app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryEstimate.js [hunk 10]
```diff
@@ -1,14 +1,14 @@
...
-const estimateRule = (estimate) => estimate > 0 ? estimate : '-';
+const estimateRule = (estimate) => estimate || '-';
 
 const CollapsedStoryEstimate = ({ estimate }) => (
   <span className='Story__estimated-value'>{estimateRule(estimate)}</span>
 );
 
 CollapsedStoryEstimate.propTypes = {
-  estimate: PropTypes.number,
+  estimate: PropTypes.any,
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryEstimate.js]: You could use it like this to accept both (empty) string and numbers.

```suggestion
  estimate: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
```

**Pull request merged**